server/src/main/java/org/apache/iotdb/db/qp/utils/WildcardsRemover.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.iotdb.db.query.expression.ResultColumn;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class WildcardsRemover {
+
+  private final ConcatPathOptimizer concatPathOptimizer;
+
+  private final int maxDeduplicatedPathNum;
+  private final int soffset;
+
+  private int offset;
+  private int limit;
+  private int consumed;
+
+  public WildcardsRemover(
+      ConcatPathOptimizer concatPathOptimizer, QueryOperator queryOperator, int fetchSize) {
+    this.concatPathOptimizer = concatPathOptimizer;
+
+    // Dataset of last query actually has only three columns, so we shouldn't limit the path num
+    // while constructing logical plan
+    // To avoid overflowing because logicalOptimize function may do maxDeduplicatedPathNum + 1, we
+    // set it to Integer.MAX_VALUE - 1
+    maxDeduplicatedPathNum =
+        queryOperator.isLastQuery()
+            ? Integer.MAX_VALUE - 1
+            : QueryResourceManager.getInstance().getMaxDeduplicatedPathNum(fetchSize);
+    soffset = queryOperator.getSeriesOffset();
+    offset = soffset;
+
+    final int slimit = queryOperator.getSeriesLimit();
+    limit = slimit == 0 || maxDeduplicatedPathNum < slimit ? maxDeduplicatedPathNum + 1 : slimit;
+
+    consumed = 0;
+  }
+
+  public WildcardsRemover(ConcatPathOptimizer concatPathOptimizer) {
+    this.concatPathOptimizer = concatPathOptimizer;
+
+    maxDeduplicatedPathNum = Integer.MAX_VALUE - 1;
+    soffset = 0;
+    limit = maxDeduplicatedPathNum + 1;
+    consumed = 0;
+  }
+
+  public List<PartialPath> removeWildcardFrom(PartialPath path) throws LogicalOptimizeException {
+    try {
+      Pair<List<PartialPath>, Integer> pair =
+          concatPathOptimizer.removeWildcard(path, limit, offset);
+
+      consumed += pair.right;
+      if (offset != 0) {
+        int delta = offset - pair.right;
+        offset = Math.max(delta, 0);
+        if (delta < 0) {
+          limit += delta;
+        }
+      } else {
+        limit -= pair.right;
+      }
+
+      return pair.left;
+    } catch (MetadataException e) {
+      throw new LogicalOptimizeException("error occurred when removing star: " + e.getMessage());
+    }
+  }
+
+  public List<List<Expression>> removeWildcardsFrom(List<Expression> expressions)
+      throws LogicalOptimizeException {
+    List<List<Expression>> extendedExpressions = new ArrayList<>();
+
+    boolean atLeastOneSeriesNotExisted = false;
+    for (Expression originExpression : expressions) {
+      List<Expression> actualExpressions = new ArrayList<>();
+      originExpression.removeWildcards(
+          new WildcardsRemover(concatPathOptimizer), actualExpressions);
+      if (actualExpressions.isEmpty()) {
+        atLeastOneSeriesNotExisted = true;
+        break;
+      }
+      extendedExpressions.add(actualExpressions);
+    }
+    if (atLeastOneSeriesNotExisted) {
+      return Collections.emptyList();
+    }
```

Comment:
Now that once one series is not existed, break the loop and return, I think we don't need the flag.
```suggestion
    for (Expression originExpression : expressions) {
      List<Expression> actualExpressions = new ArrayList<>();
      originExpression.removeWildcards(
          new WildcardsRemover(concatPathOptimizer), actualExpressions);
      if (actualExpressions.isEmpty()) {
        return Collections.emptyList();
      }
      extendedExpressions.add(actualExpressions);
    }
```

Reply:
And I doubt that whether we should return a empty list or throw an exception.

Reply:
> Now that once one series is not existed, break the loop and return, I think we don't need the flag.

Very GOOD finding! I fixed it.

>  And I doubt that whether we should return a empty list or throw an exception.

We ignore the eval of the function which has at least one non-existence series as input. 

See IOTDB-1212: https://github.com/apache/iotdb/pull/3101